src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs

```diff
@@ -0,0 +1,216 @@
+        {
+            int[][] counters;
+            int count;
+            lock (this)
+            {
+                counters = _counters;
+                count = _count;
+                _counters = new int[ExponentArraySize][];
+                _count = 0;
+            }
+            QuantileValue[] quantiles = new QuantileValue[_config.Quantiles.Length];
+            int nextQuantileIndex = 0;
+            if (nextQuantileIndex == _config.Quantiles.Length)
+            {
+                return new HistogramStatistics(quantiles);
+            }
+
+            // Reduce the count if there are any NaN or +/-Infinity values that were logged
+            count -= GetInvalidCount(counters);
+
+            // Consider each bucket to have N entries in it, and each entry has value GetBucketCanonicalValue().
+            // If all these entries were inserted in a sorted array, we are trying to find the value of the entry with
+            // index=target.
+            int target = QuantileToRank(_config.Quantiles[nextQuantileIndex], count);
+
+            // the total number of entries in all buckets iterated so far
+            int cur = 0;
+            foreach (Bucket b in IterateBuckets(counters))
+            {
+                cur += b.Count;
+                while (cur > target)
+                {
+                    quantiles[nextQuantileIndex] = new QuantileValue(
+                        _config.Quantiles[nextQuantileIndex], b.Value);
+                    nextQuantileIndex++;
+                    if (nextQuantileIndex == _config.Quantiles.Length)
+                    {
+                        return new HistogramStatistics(quantiles);
+                    }
+                    target = QuantileToRank(_config.Quantiles[nextQuantileIndex], count);
+                }
+            }
+
+            Debug.Assert(count == 0);
+#if NO_ARRAY_EMPTY_SUPPORT
+            return new HistogramStatistics(EmptyQuantiles);
+#else
+            return new HistogramStatistics(Array.Empty<QuantileValue>());
+#endif
+        }
+
+        private int GetInvalidCount(int[][] counters)
+        {
+            int[] positiveInfAndNan = counters[ExponentArraySize / 2 - 1];
+            int[] negativeInfAndNan = counters[ExponentArraySize - 1];
+            int count = 0;
+            if (positiveInfAndNan != null)
+            {
+                foreach (int bucketCount in positiveInfAndNan)
+                {
+                    count += bucketCount;
+                }
+            }
+            if (negativeInfAndNan != null)
+            {
+                foreach (int bucketCount in negativeInfAndNan)
+                {
+                    count += bucketCount;
+                }
+            }
+            return count;
+        }
+
+        private IEnumerable<Bucket> IterateBuckets(int[][] counters)
+        {
+            // iterate over the negative exponent buckets
+            const int lowestNegativeOffset = ExponentArraySize / 2;
```

Comment:
Nit: lowestNegativeOffset => LowestNegativeOffset

Reply:
just curious, do you expect it would be hard to create analyzers to enforce some of the naming and capitalization suggestions? I appreciate you taking the time to review it all but I also feel bummed that you would need to.

Reply:
(edit has been applied)

Reply:
They exist.  We're just not yet consistent enough in some of our existing code to enable it, and there's no great baselining mechanism.